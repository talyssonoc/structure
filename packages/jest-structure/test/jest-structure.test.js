const { attributes } = require('structure');
require('../extend-expect');

describe('jest-structure', () => {
  let User;

  beforeEach(() => {
    const Book = attributes({
      name: {
        type: String,
        required: true,
      },
    })(class Book {});

    User = attributes({
      name: {
        type: String,
        required: true,
        minLength: 2,
        alphanumeric: true,
      },
      age: {
        type: Number,
        required: true,
        positive: true,
      },
      favoriteBook: Book,
    })(class User {});
  });

  describe('toBeValid', () => {
    it('suceeds for valid structures', () => {
      const user = new User({ name: 'abc', age: 42 });

      expect(user).toBeValidStructure();
    });

    it('fails for invalid structures', () => {
      const user = new User();

      expect(() => {
        expect(user).toBeValidStructure();
      }).toThrowErrorMatchingSnapshot();
    });

    describe('.not', () => {
      it('succeeds for invalid structures', () => {
        const user = new User();

        expect(user).not.toBeValidStructure();
      });

      it('fails for valid structures', () => {
        const user = new User({ name: 'abc', age: 42 });

        expect(() => {
          expect(user).not.toBeValidStructure();
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('toBeInvalid', () => {
    it('suceeds for invalid structures', () => {
      const user = new User();

      expect(user).toBeInvalidStructure();
    });

    it('fails for valid structures', () => {
      const user = new User({ name: 'abc', age: 42 });

      expect(() => {
        expect(user).toBeInvalidStructure();
      }).toThrowErrorMatchingSnapshot();
    });

    describe('.not', () => {
      it('succeeds for valid structures', () => {
        const user = new User({ name: 'abc', age: 42 });

        expect(user).not.toBeInvalidStructure();
      });

      it('fails for invalid structures', () => {
        const user = new User();

        expect(() => {
          expect(user).not.toBeInvalidStructure();
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('toHaveInvalidAttribute', () => {
    it('can not be used with .not', () => {
      const user = new User();

      expect(() => {
        expect(user).not.toHaveInvalidAttribute(['name']);
      }).toThrowErrorMatchingSnapshot();
    });

    it('can not be called without attribute path', () => {
      const user = new User();

      expect(() => {
        expect(user).toHaveInvalidAttribute();
      }).toThrowErrorMatchingSnapshot();
    });

    it('can not be called with empty attribute path', () => {
      const user = new User();

      expect(() => {
        expect(user).toHaveInvalidAttribute([]);
      }).toThrowErrorMatchingSnapshot();
    });

    describe('when only attribute path is passed', () => {
      it('fails for valid structures', () => {
        const user = new User({ name: 'abc', age: 42 });

        expect(() => {
          expect(user).toHaveInvalidAttribute(['name']);
        }).toThrowErrorMatchingSnapshot();
      });

      describe('when the attribute is the only one invalid', () => {
        it('succeeds ', () => {
          const user = new User({ age: 42 });

          expect(user).toHaveInvalidAttribute(['name']);
        });
      });

      describe('when the attribute is not the only one invalid', () => {
        it('succeeds ', () => {
          const user = new User({});

          expect(user).toHaveInvalidAttribute(['name']);
        });
      });

      describe('when the attribute is not the invalid one', () => {
        it('fails ', () => {
          const user = new User({ name: 'abc' });

          expect(() => {
            expect(user).toHaveInvalidAttribute(['name']);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });

    describe('when attribute path and error messages are passed', () => {
      it('fails for valid structures', () => {
        const user = new User({ name: 'abc', age: 42 });

        expect(() => {
          expect(user).toHaveInvalidAttribute(['name'], ['nope']);
        }).toThrowErrorMatchingSnapshot();
      });

      describe('when the attribute is the only one invalid', () => {
        describe('when the errors are passed the same order returned by validate()', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: 42 });

            expect(user).toHaveInvalidAttribute(
              ['name'],
              [
                '"name" length must be at least 2 characters long',
                '"name" must only contain alpha-numeric characters',
              ]
            );
          });
        });

        describe('when the errors are passed in an order different from the returned by validate()', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: 42 });

            expect(user).toHaveInvalidAttribute(
              ['name'],
              [
                '"name" must only contain alpha-numeric characters',
                '"name" length must be at least 2 characters long',
              ]
            );
          });
        });

        describe('when the expected errors are a subset of the actual errors', () => {
          it('fails', () => {
            const user = new User({ name: '$', age: 42 });

            expect(() => {
              expect(user).toHaveInvalidAttribute(
                ['name'],
                ['"name" length must be at least 2 characters long']
              );
            }).toThrowErrorMatchingSnapshot();
          });
        });

        describe('when the expected errors are a superset of the actual errors', () => {
          it('fails', () => {
            const user = new User({ name: '$', age: 42 });

            expect(() => {
              expect(user).toHaveInvalidAttribute(
                ['name'],
                [
                  '"name" must only contain alpha-numeric characters',
                  '"name" is not from this planet',
                  '"name" length must be at least 2 characters long',
                ]
              );
            }).toThrowErrorMatchingSnapshot();
          });
        });

        describe('when using arrayContaining', () => {
          describe('when arrayContaining is a subset of the errors', () => {
            it('succeeds', () => {
              const user = new User({ name: '$', age: 42 });

              expect(user).toHaveInvalidAttribute(
                ['name'],
                expect.arrayContaining(['"name" length must be at least 2 characters long'])
              );
            });
          });

          describe('when arrayContaining is a superset of the errors', () => {
            it('fails', () => {
              const user = new User({ name: '$', age: 42 });

              expect(() => {
                expect(user).toHaveInvalidAttribute(
                  ['name'],
                  expect.arrayContaining([
                    '"name" must only contain alpha-numeric characters',
                    '"name" is not from this planet',
                    '"name" length must be at least 2 characters long',
                  ])
                );
              }).toThrowErrorMatchingSnapshot();
            });
          });

          describe('when arrayContaining has an intersection with the errors + other different errors', () => {
            it('fails', () => {
              const user = new User({ name: '$', age: 42 });

              expect(() => {
                expect(user).toHaveInvalidAttribute(
                  ['name'],
                  expect.arrayContaining([
                    '"name" is not from this planet',
                    '"name" length must be at least 2 characters long',
                  ])
                );
              }).toThrowErrorMatchingSnapshot();
            });
          });
        });
      });

      describe('when the attribute is not the only one invalid', () => {
        it('succeeds ', () => {
          const user = new User({});

          expect(user).toHaveInvalidAttribute(['name'], ['"name" is required']);
        });
      });

      describe('when the attribute is not the invalid one', () => {
        it('fails ', () => {
          const user = new User({ age: 42 });

          expect(() => {
            expect(user).toHaveInvalidAttribute(['age'], ['"age" is wrong']);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });

  describe('toHaveInvalidAttributes', () => {
    it('can not be used with .not', () => {
      const user = new User();

      expect(() => {
        expect(user).not.toHaveInvalidAttributes([]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('can not be called without the array of expected errors', () => {
      const user = new User();

      expect(() => {
        expect(user).toHaveInvalidAttributes();
      }).toThrowErrorMatchingSnapshot();
    });

    it('can not be called with empty array of expected errors', () => {
      const user = new User();

      expect(() => {
        expect(user).toHaveInvalidAttributes([]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('can not be called with no path', () => {
      const user = new User({ name: '$', age: -42 });

      expect(() => {
        expect(user).toHaveInvalidAttributes([{ path: [] }, { path: ['age'] }]);
      }).toThrowErrorMatchingSnapshot();
    });

    describe('when only paths are passed', () => {
      it('fails for valid structures', () => {
        const user = new User({ name: 'abc', age: 42 });

        expect(() => {
          expect(user).toHaveInvalidAttributes([{ path: ['name'] }]);
        }).toThrowErrorMatchingSnapshot();
      });

      describe('when all the passed attributes are invalid', () => {
        describe('when attributes are passed in the same order returned by validate', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: -42 });

            expect(user).toHaveInvalidAttributes([{ path: ['name'] }, { path: ['age'] }]);
          });
        });

        describe('when attributes are passed in an order different from the returned by validate', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: -42 });

            expect(user).toHaveInvalidAttributes([{ path: ['age'] }, { path: ['name'] }]);
          });
        });
      });

      describe('when only some of the passed attributes are invalid', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: 42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([{ path: ['age'] }, { path: ['name'] }]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when none of the passed attributes are invalid', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: 42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([{ path: ['age'] }]);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });

    describe('when paths and messages are passed', () => {
      it('fails for valid structures', () => {
        const user = new User({ name: 'abc', age: 42 });

        expect(() => {
          expect(user).toHaveInvalidAttributes([
            { path: ['name'], messages: ['something wrong is not right'] },
          ]);
        }).toThrowErrorMatchingSnapshot();
      });

      describe('when attribute is invalid and messages are right', () => {
        describe('when messages are passed in the same order returned by validate', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: 42 });

            expect(user).toHaveInvalidAttributes([
              {
                path: ['name'],
                messages: [
                  '"name" length must be at least 2 characters long',
                  '"name" must only contain alpha-numeric characters',
                ],
              },
            ]);
          });
        });

        describe('when messages are passed in a order different from the returned by validate', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: 42 });

            expect(user).toHaveInvalidAttributes([
              {
                path: ['name'],
                messages: [
                  '"name" must only contain alpha-numeric characters',
                  '"name" length must be at least 2 characters long',
                ],
              },
            ]);
          });
        });
      });

      describe('when attribute is invalid but messages array is empty', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: 42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([{ path: ['name'], messages: [] }]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when there are missing attributes', () => {
        it('fails', () => {
          const user = new User({ name: 'A', age: -42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              { path: ['name'], messages: ['"name" length must be at least 2 characters long'] },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when there are missing messages', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: 42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              { path: ['name'], messages: ['"name" length must be at least 2 characters long'] },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when there are missing attributes and messages', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: -42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              { path: ['name'], messages: ['"name" length must be at least 2 characters long'] },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });

    describe('when messages are passed only for some', () => {
      it('fails for valid structures', () => {
        const user = new User({ name: 'abc', age: 42 });

        expect(() => {
          expect(user).toHaveInvalidAttributes([
            { path: ['name'], messages: ['something wrong is not right'] },
            { path: ['age'] },
          ]);
        }).toThrowErrorMatchingSnapshot();
      });

      describe('when attribute is invalid and messages are right', () => {
        describe('when messages are passed in the same order returned by validate', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: -42 });

            expect(user).toHaveInvalidAttributes([
              {
                path: ['name'],
                messages: [
                  '"name" length must be at least 2 characters long',
                  '"name" must only contain alpha-numeric characters',
                ],
              },
              { path: ['age'] },
            ]);
          });
        });

        describe('when messages are passed in a order different from the returned by validate', () => {
          it('succeeds', () => {
            const user = new User({ name: '$', age: -42 });

            expect(user).toHaveInvalidAttributes([
              {
                path: ['name'],
                messages: [
                  '"name" must only contain alpha-numeric characters',
                  '"name" length must be at least 2 characters long',
                ],
              },
              { path: ['age'] },
            ]);
          });
        });
      });

      describe('when attribute is invalid but messages array is empty', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: -42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              { path: ['name'], messages: [] },
              { path: ['age'] },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when there are missing attributes', () => {
        it('fails', () => {
          const user = new User({ name: 'A', age: -42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([{ path: ['name'] }]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when there are missing messages', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: -42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              { path: ['name'], messages: ['"name" length must be at least 2 characters long'] },
              { path: ['age'] },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when there are missing attributes and messages', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: -42, favoriteBook: {} });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              { path: ['name'], messages: ['"name" length must be at least 2 characters long'] },
              { path: ['age'] },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });

    describe('when using arrayContaining', () => {
      describe('when arrayContaining is a subset of the errors', () => {
        it('succeeds', () => {
          const user = new User({ name: '$', age: 42 });

          expect(user).toHaveInvalidAttributes([
            {
              path: ['name'],
              messages: expect.arrayContaining([
                '"name" length must be at least 2 characters long',
              ]),
            },
          ]);
        });
      });

      describe('when arrayContaining is a superset of the errors', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: 42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              {
                path: ['name'],
                messages: expect.arrayContaining([
                  '"name" must only contain alpha-numeric characters',
                  '"name" is not from this planet',
                  '"name" length must be at least 2 characters long',
                ]),
              },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });

      describe('when arrayContaining has an intersection with the errors + other different errors', () => {
        it('fails', () => {
          const user = new User({ name: '$', age: 42 });

          expect(() => {
            expect(user).toHaveInvalidAttributes([
              {
                path: ['name'],
                messages: expect.arrayContaining([
                  '"name" is not from this planet',
                  '"name" length must be at least 2 characters long',
                ]),
              },
            ]);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });
});
