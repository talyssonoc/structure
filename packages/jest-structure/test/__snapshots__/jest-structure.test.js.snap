// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-structure toBeInvalid .not fails for invalid structures 1`] = `
"Expected: to be [32mvalid[39m
Received: is [31minvalid[39m"
`;

exports[`jest-structure toBeInvalid fails for valid structures 1`] = `
"Expected: to be [32minvalid[39m
Received: is [31mvalid[39m"
`;

exports[`jest-structure toBeValid .not fails for valid structures 1`] = `
"Expected: to be [32minvalid[39m
Received: is [31mvalid[39m"
`;

exports[`jest-structure toBeValid fails for invalid structures 1`] = `
"Expected: to be [32mvalid[39m
Received: is [31minvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttribute can not be called with empty attribute path 1`] = `
"toHaveInvalidAttribute must not be called without the attribute path
Example: [2mexpect([22m[31mstructure[39m[2m).[22mtoHaveInvalidAttribute[2m([22m[32mattributePath[39m[2m, [22m[32m[errorMessages][39m[2m)[22m"
`;

exports[`jest-structure toHaveInvalidAttribute can not be called without attribute path 1`] = `
"toHaveInvalidAttribute must not be called without the attribute path
Example: [2mexpect([22m[31mstructure[39m[2m).[22mtoHaveInvalidAttribute[2m([22m[32mattributePath[39m[2m, [22m[32m[errorMessages][39m[2m)[22m"
`;

exports[`jest-structure toHaveInvalidAttribute can not be used with .not 1`] = `"toHaveInvalidAttribute must not be used with .not"`;

exports[`jest-structure toHaveInvalidAttribute when attribute path and error messages are passed fails for valid structures 1`] = `
"Expected: to be [32minvalid[39m
Received: is [31mvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttribute when attribute path and error messages are passed when the attribute is not the invalid one fails  1`] = `
"Expected: age to be [32minvalid[39m
Received: age is [31mvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttribute when attribute path and error messages are passed when the attribute is the only one invalid when the expected errors are a subset of the actual errors fails 1`] = `
"[32m- Expected name error messages  - 0[39m
[31m+ Received name error messages  + 1[39m

[2m  Array [[22m
[2m    \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[31m+   \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttribute when attribute path and error messages are passed when the attribute is the only one invalid when the expected errors are a superset of the actual errors fails 1`] = `
"[32m- Expected name error messages  - 1[39m
[31m+ Received name error messages  + 0[39m

[2m  Array [[22m
[2m    \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[22m
[32m-   \\"\\\\\\"name\\\\\\" is not from this planet\\",[39m
[2m    \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttribute when attribute path and error messages are passed when the attribute is the only one invalid when using arrayContaining when arrayContaining has an intersection with the errors + other different errors fails 1`] = `
"Expected name error messages: [32mArrayContaining [\\"\\\\\\"name\\\\\\" is not from this planet\\", \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\"][39m
Received name error messages: [31m[\\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\", \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\"][39m"
`;

exports[`jest-structure toHaveInvalidAttribute when attribute path and error messages are passed when the attribute is the only one invalid when using arrayContaining when arrayContaining is a superset of the errors fails 1`] = `
"Expected name error messages: [32mArrayContaining [\\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\", \\"\\\\\\"name\\\\\\" is not from this planet\\", \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\"][39m
Received name error messages: [31m[\\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\", \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\"][39m"
`;

exports[`jest-structure toHaveInvalidAttribute when only attribute path is passed fails for valid structures 1`] = `
"Expected: to be [32minvalid[39m
Received: is [31mvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttribute when only attribute path is passed when the attribute is not the invalid one fails  1`] = `
"Expected: name to be [32minvalid[39m
Received: name is [31mvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttributes can not be called with empty array of expected errors 1`] = `
"toHaveInvalidAttributes must not be called without the expected errros
Example: [2mexpect([22m[31mstructure[39m[2m).[22mtoHaveInvalidAttributes[2m([22m[32m[{ path (required), messages (optional) }][39m[2m)[22m"
`;

exports[`jest-structure toHaveInvalidAttributes can not be called with no path 1`] = `
"toHaveInvalidAttributes must not be called without the attribute paths
Example: [2mexpect([22m[31mstructure[39m[2m).[22mtoHaveInvalidAttributes[2m([22m[32m[{ path (required), messages (optional) }][39m[2m)[22m"
`;

exports[`jest-structure toHaveInvalidAttributes can not be called without the array of expected errors 1`] = `
"toHaveInvalidAttributes must not be called without the expected errros
Example: [2mexpect([22m[31mstructure[39m[2m).[22mtoHaveInvalidAttributes[2m([22m[32m[{ path (required), messages (optional) }][39m[2m)[22m"
`;

exports[`jest-structure toHaveInvalidAttributes can not be used with .not 1`] = `"toHaveInvalidAttributes must not be used with .not"`;

exports[`jest-structure toHaveInvalidAttributes when messages are passed only for some fails for valid structures 1`] = `
"Expected: to be [32minvalid[39m
Received: is [31mvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttributes when messages are passed only for some when attribute is invalid but messages array is empty fails 1`] = `
"[32m- Expected errors  - 1[39m
[31m+ Received errors  + 4[39m

[33m@@ -1,8 +1,11 @@[39m
[2m  Array [[22m
[2m    Object {[22m
[32m-     \\"messages\\": Array [],[39m
[31m+     \\"messages\\": Array [[39m
[31m+       \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[39m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[31m+     ],[39m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m
[2m    Object {[22m"
`;

exports[`jest-structure toHaveInvalidAttributes when messages are passed only for some when there are missing attributes and messages fails 1`] = `
"[32m- Expected errors  -  0[39m
[31m+ Received errors  + 10[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"messages\\": Array [[22m
[2m        \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[2m      ],[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m
[2m    Object {[22m
[2m      \\"path\\": Array [[22m
[2m        \\"age\\",[22m
[31m+     ],[39m
[31m+   },[39m
[31m+   Object {[39m
[31m+     \\"messages\\": Array [[39m
[31m+       \\"\\\\\\"favoriteBook.name\\\\\\" is required\\",[39m
[31m+     ],[39m
[31m+     \\"path\\": Array [[39m
[31m+       \\"favoriteBook\\",[39m
[31m+       \\"name\\",[39m
[2m      ],[22m
[2m    },[22m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttributes when messages are passed only for some when there are missing attributes fails 1`] = `
"[32m- Expected errors  - 0[39m
[31m+ Received errors  + 8[39m

[33m@@ -2,6 +2,14 @@[39m
[2m    Object {[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m
[31m+   Object {[39m
[31m+     \\"messages\\": Array [[39m
[31m+       \\"\\\\\\"age\\\\\\" must be a positive number\\",[39m
[31m+     ],[39m
[31m+     \\"path\\": Array [[39m
[31m+       \\"age\\",[39m
[31m+     ],[39m
[31m+   },[39m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttributes when messages are passed only for some when there are missing messages fails 1`] = `
"[32m- Expected errors  - 0[39m
[31m+ Received errors  + 1[39m

[33m@@ -1,9 +1,10 @@[39m
[2m  Array [[22m
[2m    Object {[22m
[2m      \\"messages\\": Array [[22m
[2m        \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[2m      ],[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m"
`;

exports[`jest-structure toHaveInvalidAttributes when only paths are passed fails for valid structures 1`] = `
"Expected: to be [32minvalid[39m
Received: is [31mvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttributes when only paths are passed when none of the passed attributes are invalid fails 1`] = `
"[32m- Expected errors  - 1[39m
[31m+ Received errors  + 5[39m

[2m  Array [[22m
[2m    Object {[22m
[31m+     \\"messages\\": Array [[39m
[31m+       \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[39m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[31m+     ],[39m
[2m      \\"path\\": Array [[22m
[32m-       \\"age\\",[39m
[31m+       \\"name\\",[39m
[2m      ],[22m
[2m    },[22m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttributes when only paths are passed when only some of the passed attributes are invalid fails 1`] = `
"[32m- Expected errors  - 5[39m
[31m+ Received errors  + 0[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"path\\": Array [[22m
[32m-       \\"age\\",[39m
[32m-     ],[39m
[32m-   },[39m
[32m-   Object {[39m
[32m-     \\"path\\": Array [[39m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttributes when paths and messages are passed fails for valid structures 1`] = `
"Expected: to be [32minvalid[39m
Received: is [31mvalid[39m"
`;

exports[`jest-structure toHaveInvalidAttributes when paths and messages are passed when attribute is invalid but messages array is empty fails 1`] = `
"[32m- Expected errors  - 1[39m
[31m+ Received errors  + 4[39m

[2m  Array [[22m
[2m    Object {[22m
[32m-     \\"messages\\": Array [],[39m
[31m+     \\"messages\\": Array [[39m
[31m+       \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[39m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[31m+     ],[39m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttributes when paths and messages are passed when there are missing attributes and messages fails 1`] = `
"[32m- Expected errors  - 0[39m
[31m+ Received errors  + 9[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"messages\\": Array [[22m
[2m        \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[2m      ],[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[31m+     ],[39m
[31m+   },[39m
[31m+   Object {[39m
[31m+     \\"messages\\": Array [[39m
[31m+       \\"\\\\\\"age\\\\\\" must be a positive number\\",[39m
[31m+     ],[39m
[31m+     \\"path\\": Array [[39m
[31m+       \\"age\\",[39m
[2m      ],[22m
[2m    },[22m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttributes when paths and messages are passed when there are missing attributes fails 1`] = `
"[32m- Expected errors  - 0[39m
[31m+ Received errors  + 8[39m

[33m@@ -5,6 +5,14 @@[39m
[2m      ],[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m
[31m+   Object {[39m
[31m+     \\"messages\\": Array [[39m
[31m+       \\"\\\\\\"age\\\\\\" must be a positive number\\",[39m
[31m+     ],[39m
[31m+     \\"path\\": Array [[39m
[31m+       \\"age\\",[39m
[31m+     ],[39m
[31m+   },[39m
[2m  ][22m"
`;

exports[`jest-structure toHaveInvalidAttributes when paths and messages are passed when there are missing messages fails 1`] = `
"[32m- Expected errors  - 0[39m
[31m+ Received errors  + 1[39m

[33m@@ -1,9 +1,10 @@[39m
[2m  Array [[22m
[2m    Object {[22m
[2m      \\"messages\\": Array [[22m
[2m        \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[2m      ],[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m"
`;

exports[`jest-structure toHaveInvalidAttributes when using arrayContaining when arrayContaining has an intersection with the errors + other different errors fails 1`] = `
"[32m- Expected errors  - 2[39m
[31m+ Received errors  + 2[39m

[33m@@ -1,10 +1,10 @@[39m
[2m  Array [[22m
[2m    Object {[22m
[32m-     \\"messages\\": ArrayContaining [[39m
[32m-       \\"\\\\\\"name\\\\\\" is not from this planet\\",[39m
[31m+     \\"messages\\": Array [[39m
[2m        \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[31m+       \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[39m
[2m      ],[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m
[2m    },[22m"
`;

exports[`jest-structure toHaveInvalidAttributes when using arrayContaining when arrayContaining is a superset of the errors fails 1`] = `
"[32m- Expected errors  - 2[39m
[31m+ Received errors  + 1[39m

[33m@@ -1,10 +1,9 @@[39m
[2m  Array [[22m
[2m    Object {[22m
[32m-     \\"messages\\": ArrayContaining [[39m
[31m+     \\"messages\\": Array [[39m
[2m        \\"\\\\\\"name\\\\\\" must only contain alpha-numeric characters\\",[22m
[32m-       \\"\\\\\\"name\\\\\\" is not from this planet\\",[39m
[2m        \\"\\\\\\"name\\\\\\" length must be at least 2 characters long\\",[22m
[2m      ],[22m
[2m      \\"path\\": Array [[22m
[2m        \\"name\\",[22m
[2m      ],[22m"
`;
